todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt

```diff
@@ -0,0 +1,114 @@
+import com.example.android.architecture.blueprints.todoapp.addedittask.domain.AddEditTaskModel
+import com.example.android.architecture.blueprints.todoapp.addedittask.domain.AddEditTaskModel.Mode.Add
+import com.example.android.architecture.blueprints.todoapp.addedittask.domain.AddEditTaskModel.Mode.Edit
+import com.example.android.architecture.blueprints.todoapp.addedittask.domain.Model
+import com.example.android.architecture.blueprints.todoapp.addedittask.effecthandlers.createEffectHandlers
+import com.example.android.architecture.blueprints.todoapp.addedittask.view.AddEditTaskViews
+import com.example.android.architecture.blueprints.todoapp.addedittask.view.toAddEditTaskModel
+import com.example.android.architecture.blueprints.todoapp.addedittask.view.toBundle
+import com.example.android.architecture.blueprints.todoapp.data.Task
+import com.example.android.architecture.blueprints.todoapp.data.TaskBundlePacker
+import com.spotify.mobius.MobiusLoop
+import io.reactivex.functions.Action
+
+/** Main UI for the add task screen. Users can enter a task title and description.  */
+class AddEditTaskFragment : Fragment() {
+
+    private lateinit var mController: MobiusLoop.Controller<AddEditTaskModel, AddEditTaskEvent>
+
+    override fun onCreateView(
+            inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+
+        val fab = activity!!.findViewById<FloatingActionButton>(R.id.fab_edit_task_done)
+        val views = AddEditTaskViews(inflater, container!!, fab)
+
+        mController = createController(
+                createEffectHandlers(context!!, Action { finishWithResultOk() }, Action { views.showEmptyTaskError() }),
+                resolveDefaultModel(savedInstanceState))
+        mController.connect(views)
+
+        setHasOptionsMenu(true)
+        return views.rootView
+    }
+
+    private fun resolveDefaultModel(savedInstanceState: Bundle?): Model {
+        if (savedInstanceState != null && savedInstanceState.containsKey(ADD_EDIT_TASK_MODEL_RESTORE_KEY)) {
+            return savedInstanceState.getBundle(ADD_EDIT_TASK_MODEL_RESTORE_KEY).toAddEditTaskModel()
+        }
+
+        val arguments = arguments
+        if (arguments != null && arguments.containsKey(TASK_ARGUMENT)) {
+            val (id, details) = TaskBundlePacker.taskFromBundle(arguments.getBundle(TASK_ARGUMENT))
+            return Model(details = details, mode = Edit(id))
+        }
+
+        return Model(mode = Add)
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putBundle(ADD_EDIT_TASK_MODEL_RESTORE_KEY, mController.model.toBundle())
+    }
+
+    override fun onResume() {
+        super.onResume()
+        mController.start()
+    }
+
+    override fun onPause() {
+        mController.stop()
+        super.onPause()
+    }
+
+    private fun finishWithResultOk() {
+        activity!!.setResult(Activity.RESULT_OK)
```

Comment:
I generally prefer ? for these types of use cases.

Reply:
that has a differrent meaning though - if the assumption is that this never will be called when activity is `null`, then it makes sense to use `!!` and force an NPE instead of silently dropping it. and ofc it'd be even nicer if the code was restructured in such a way that the activity never is null where this code lives.